crates/turborepo-filewatch/src/globwatcher.rs

```diff
@@ -0,0 +1,572 @@
+                // Build a set of candidate globs that *may* have changed.
+                // An empty set translates to all globs have not changed.
+                if let Some(unchanged_globs) = self.hash_globs.get(&hash) {
+                    candidates.retain(|glob_str| {
+                        // We are keeping the globs from candidates that
+                        // we don't have a record of as unchanged.
+                        // If we do have a record, drop it from candidates.
+                        !unchanged_globs.include.contains_key(glob_str)
+                    });
+                }
+                // If the client has gone away, we don't care about the error
+                let _ = resp.send(Ok(candidates));
+            }
+        }
+    }
+
+    fn handle_file_event(
+        &mut self,
+        file_event: Result<Result<Event, NotifyError>, broadcast::error::RecvError>,
+    ) {
+        match file_event {
+            Err(broadcast::error::RecvError::Closed) => (),
+            Err(e @ broadcast::error::RecvError::Lagged(_)) => self.on_error(e.into()),
+            Ok(Err(error)) => self.on_error(error.into()),
+            Ok(Ok(file_event)) => {
+                for path in file_event.paths {
+                    let path = AbsoluteSystemPathBuf::try_from(path)
+                        .expect("filewatching should produce absolute paths");
+                    let Ok(to_match) = self.root.anchor(path) else {
+                        // irrelevant filesystem update
+                        return;
+                    };
+                    self.handle_path_change(&to_match.to_unix());
+                }
+            }
+        }
+    }
+
+    async fn watch(mut self) {
+        loop {
+            tokio::select! {
+                _ = &mut self.exit_signal => return,
+                Some(query) = self.query_recv.recv().into_future() => self.handle_query(query),
+                file_event = self.recv.recv().into_future() => self.handle_file_event(file_event)
+            }
+        }
+    }
+
+    /// on_error takes the conservative approach of considering everything
+    /// changed in the event of any error related to filewatching
+    fn on_error(&mut self, err: WatchError) {
+        warn!(
+            "encountered filewatching error, flushing all globs: {}",
+            err
+        );
+        self.hash_globs.clear();
+        self.glob_statuses.clear();
+    }
+
+    fn handle_path_change(&mut self, path: &RelativeUnixPath) {
+        self.glob_statuses.retain(|glob_str, (glob, hash_globs)| {
+            // If this is not a match, we aren't modifying this glob, bail early and mark
+            // for retention.
+            if !glob.is_match(path) {
+                return true;
+            }
+            // We have a match. Check which hashes need invalidation.
+            hash_globs.retain(|hash| {
+                let Some(glob_set) = self.hash_globs.get_mut(hash) else {
```

Comment:
If the hash isn't in `self.hash_globs`, doesn't that mean we're not watching any globs for this hash? Shouldn't we the hash from `hash_globs`

Reply:
Also side note, it's a little confusing using `hash_globs` for both the `HashSet<Hash>` and the `HashMap<Hash, GlobSet>`

Reply:
Agreed. Renamed one (the value for `glob_hashes`) to `hashes_for_glob`, which better describes what we're iterating over.

Reply:
I think the scenario where we don't find the hash should never happen? But probably we should remove the hash that we couldn't find.